### PR TITLE
#4320 : Wrong column name when double uppercase

### LINKED ISSUE #4320: Wrong column name in case of double uppercase
##### **Overview of the issue**

Jhipster generator create a wrong liquibase changelog when we have a double Uppercase name.
And we have an error on unit test:

`org.hibernate.HibernateException: Missing column: socrevision_id in SERVER.PUBLIC.FILE`

I have a relation in my uml : soCRevision.

According to SpringNamingStrategy:

```
protected static String addUnderscores(String name) {
		StringBuilder buf = new StringBuilder( name.replace('.', '_') );
		for (int i=1; i<buf.length()-1; i++) {
			if (
				Character.isLowerCase( buf.charAt(i-1) ) &&
				Character.isUpperCase( buf.charAt(i) ) &&
				Character.isLowerCase( buf.charAt(i+1) )
			) {
				buf.insert(i++, '_');
			}
		}
		return buf.toString().toLowerCase(Locale.ROOT);
	}
```

And in my liquibase:

```
<column name="so_c_revision_id" type="bigint">
                <constraints nullable="true" />
            </column>
```

##### **Motivation for or Use Case**

because unit test doesn't pass. ;o) 

##### **Reproduce the error**

create the project with the following files

##### **Suggest a Fix**
I think the problem is here:
https://github.com/jhipster/generator-jhipster/blob/01c01f63bb3206526cf92a34800475265a5e7e0d/generators/generator-base.js#L1164
```
Generator.prototype.getColumnName = function (value) {
    return _.snakeCase(value).toLowerCase();
};
```
must be replace by the Spring version:
```
protected static String addUnderscores(String name) {
		StringBuilder buf = new StringBuilder( name.replace('.', '_') );
		for (int i=1; i<buf.length()-1; i++) {
			if (
				Character.isLowerCase( buf.charAt(i-1) ) &&
				Character.isUpperCase( buf.charAt(i) ) &&
				Character.isLowerCase( buf.charAt(i+1) )
			) {
				buf.insert(i++, '_');
			}
		}
		return buf.toString().toLowerCase(Locale.ROOT);
	}
```

##### **JHipster Version(s)**

3.8.0

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

```
{
  "generator-jhipster": {
    "jhipsterVersion": "3.8.0",
    "baseName": "server",
    "packageName": "org.duff",
    "packageFolder": "org/duff",
    "serverPort": "8080",
    "authenticationType": "session",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "postgresql",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "rememberMeKey": "b74f0cc9342da628bc7064b2af4d6f48786c4f31",
    "useSass": false,
    "applicationType": "monolith",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ]
  }
}
```

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**
File.json
```
{
  "relationships": [
    {
      "relationshipType": "many-to-one",
      "relationshipName": "soCRevision",
      "otherEntityName": "revision",
      "otherEntityField": "id"
    }
  ],
  "fields": [
    {
      "fieldName": "description",
      "fieldType": "String"
    },
    {
      "fieldName": "version",
      "fieldType": "String"
    }
  ],
  "changelogDate": "20161004125225",
  "dto": "no",
  "pagination": "no",
  "service": "no",
  "entityTableName": "file"
}
```
Revision.json
```
{
  "relationships": [
    {
      "relationshipType": "one-to-many",
      "relationshipName": "soCRepository",
      "otherEntityName": "file",
      "otherEntityRelationshipName": "soCRevision"
    }
  ],
  "fields": [
   {
      "fieldName": "description",
      "fieldType": "String"
    }
  ],
  "changelogDate": "20161004125239",
  "dto": "no",
  "pagination": "no",
  "service": "no",
  "entityTableName": "revision"
}
```

##### **Browsers and Operating System**

Windows, Chrome and Firfox

### PR DESCRIPTION
**Pull request by Dufgui**
fix #4320

### CODE REVIEW

File: generators/generator-base.js [hunk 1]
```diff
@@ -1161,7 +1161,26 @@ Generator.prototype.getTableName = function (value) {
  * @param {string} value - table column name string
  */
 Generator.prototype.getColumnName = function (value) {
-    return _.snakeCase(value).toLowerCase();
+    value = value.replace('.', '_');
```
Comment (deepu105) [inline generators/generator-base.js]: I dont think it will ever have dot in the passed value

File: generators/generator-base.js [hunk 2]
```diff
@@ -1161,7 +1161,26 @@ Generator.prototype.getTableName = function (value) {
...
+    var res = '';
+    if(value && value.length > 0)
+        res = value[0];
```
Comment (deepu105) [inline generators/generator-base.js]: just return and stop flow here

File: generators/generator-base.js [hunk 3]
```diff
@@ -1161,7 +1161,26 @@ Generator.prototype.getTableName = function (value) {
...
+            value[i] !== value[i].toLowerCase() &&
+            value[i+1] !== value[i+1].toUpperCase()
+        ) {
+            res += '_' + value[i];
+        } else {
+            res += value[i];
+        }
+    }
+
+    if(value && value.length > 0)
```
Comment (deepu105) [inline generators/generator-base.js]: redundant

**Comment by jdubois** [comment]
Oh, thanks for the review @deepu105 !!

**Comment by deepu105** [comment]
I had a closer look and indeed its annoying why cant the spring people do a normal snake case 😠 
Anyways to make the solution global I would prefer the below
Add the below method to `generator-base.js` after line 1492 (private methods)
```
Generator.prototype.hibernateSnakeCase = function (value) {
    let res = '';
    if (value) {
        value = value.replace('.', '_');
        res = value[0];
        for (var i = 1, len = value.length - 1; i < len; i++) {
            if (value[i-1] !== value[i-1].toUpperCase() &&
                value[i] !== value[i].toLowerCase() &&
                value[i+1] !== value[i+1].toUpperCase()
            ) {
                res += '_' + value[i];
            } else {
                res += value[i];
            }
        }
        res += value[value.length -1];
        res = res.toLowerCase();
    }
    return res;
};
```

now change the `getColumnName` and `getTableName` methods in same file to call `this.hibernateSnakeCase(value);` instead of `_.snakeCase(value).toLowerCase();`

**Comment by Dufgui** [comment]
Ok i have checked, the spring naming strategy is the same for column and table name. So ok

**Comment by Dufgui** [comment]
So i have tested it before push and it's working.

**Comment by sendilkumarn** [comment]
@deepu105 no other api level method available to handle this? 

Is this even a standard?

**Comment by deepu105** [comment]
@sendilkumarn i think it comes from either JPA or Hibernate so we have to follow

**Comment by MathieuAA** [comment]
@deepu105 @Dufgui: With this implementation, when passing `EntityA` it returns `entitya`. By the way, where are the tests?

**Comment by deepu105** [comment]
Tests should in test-generator-base.js

Thanks & regards,
Deepu

**Pull request merged**